### PR TITLE
Refine Item Searching and add MP functionality

### PR DESCRIPTION
**Pull request by kphoenix137**
Features:
`--mp`: Allows the user to utilize multiplayer game logic
`--scanner item`: Functionally the same as `--scanner puzzler`, but renamed for this new use case. We now use this to search for any item based on the item name. Note: This doesn't filter by base item in any capacity.
`--targetStr <string>`: Used with `--scanner item` to specify the exact item name (case sensitive).

This PR also makes some minor changes to the code without affecting the function of the program.

### CODE REVIEW

File: Source/mapGen/configuration.h [hunk 1]
```diff
@@ -17,5 +17,6 @@ struct Configuration {
 	bool asciiLevels = false;
 	bool exportLevels = false;
 	std::optional<uint32_t> target = std::nullopt;
+	std::optional<const char *> targetStr = std::nullopt;
```
Comment (AJenbo) [inline Source/mapGen/configuration.h]: probably use std::string instead of const char *

File: Source/analyzer/item.cpp [hunk 2]
```diff
@@ -1,176 +1,182 @@
...
+
+		if (leveltype > 1)
+			CreateTypeItem(object[oid]._ox, object[oid]._oy, TRUE, weaponType, IMISC_NONE, TRUE, FALSE);
+		else
+			CreateTypeItem(object[oid]._ox, object[oid]._oy, FALSE, weaponType, IMISC_NONE, TRUE, FALSE);
+	} break;
+	}
+}
+
+static void DropAllItems()
```
Comment (AJenbo) [inline Source/analyzer/item.cpp:105]: why the static?

File: Source/analyzer/item.cpp [hunk 3]
```diff
@@ -1,176 +1,182 @@
...
+{
+	DropAllItems();
+
+	POI = { -1, -1 };
+
+	for (int i = 0; i < numitems; i++) {
+		int ii = itemactive[i];
+		ItemStruct &searchItem = item[ii];
+
+		if (strcmp(*Config.targetStr, searchItem._iIName) == 0) {
```
Comment (AJenbo) [inline Source/analyzer/item.cpp]: std::[string](https://cplusplus.com/reference/string/string/)::compare

**Comment by AJenbo** [comment]
Nice this should allow @Matthew-petroff and others to search for given items on floors in the candidate levels to filter out more ideal seeds to speedrun.

File: Source/analyzer/item.cpp (hunk 2 above)
Comment (kphoenix137) [reply Source/analyzer/item.cpp:105]: VS suggested it can be made static

Comment (AJenbo) [reply Source/analyzer/item.cpp:105]: Arh, thought we already had an anon name space there, but guess not

File: Source/analyzer/item.cpp [hunk 4]
```diff
@@ -122,7 +122,7 @@ static void DropAllItems()
 
 bool ScannerItem::skipLevel(int level)
 {
-	if (level == 0)
+	if (*Config.target == 0 && level >= 0 && level <= 16)
```
Comment (AJenbo) [inline Source/analyzer/item.cpp:125]: If looking for town and current level is validt? Should there have been an and somewhere? Will this become an issue when hellfire support might be added?

File: Source/funkMapGen.cpp [hunk 5]
```diff
@@ -120,7 +120,11 @@ void InitTriggers()
 		InitL3Triggers();
 	else if (leveltype == DTYPE_HELL)
 		InitL4Triggers();
+	if (currlevel == 16)
+		numtrigs--;
 	Freeupstairs();
+	if (currlevel == 16)
```
Comment (AJenbo) [inline Source/funkMapGen.cpp:126]: This was already fixed in a better way

**Comment by AJenbo** [comment]
Merged via https://github.com/Matthew-petroff/diablo-mapgen/commit/c86aa98eca1e82649232b783e24c0dee86186b7f

**Pull request closed without merging**